sycl/doc/Assert.md

```diff
@@ -0,0 +1,421 @@
+
+As usual, device-side assertions can be disabled by defining `NDEBUG` macro at
+compile time.
+
+## Use-case example
+
+```c++
+#include <cassert>
+#include <sycl/sycl.hpp>
+
+using namespace sycl;
+
+void user_func(item<2> Item) {
+  assert((Item[0] % 2) && “Nil”);
+}
+
+int main() {
+  queue Q;
+  Q.submit([&] (handler& CGH) {
+    CGH.parallel_for<class TheKernel>(range<2>{N, M}, [=](item<2> It) {
+      do_smth();
+      user_func(It);
+      do_smth_else();
+    });
+  });
+  Q.wait();
+  std::cout << “One shouldn’t see this message.“;
+  return 0;
+}
+```
+
+In this use-case every work-item with even X dimension will trigger assertion
+failure. Assertion failure should be trigger a call to `std::abort()` at host as
+described in
+[extension](extensions/Assert/SYCL_INTEL_ASSERT.asciidoc).
+Even though multiple failures of the same or different assertions can happen in
+multiple workitems, implementation is required to deliver at least one
+assertion. The assertion failure message is printed to `stderr` by DPCPP
+Runtime.
+
+When multiple kernels are enqueued and more than one fail at assertion, at least
+single assertion should be reported.
+
+
+## User requirements
+
+From user's point of view there are the following requirements:
+
+| # | Title | Description | Importance |
+| - | ----- | ----------- | ---------- |
+| 1 | Abort DPC++ application | Abort host application when assert function is called and print a message about assertion | Must have |
+| 2 | Print assert message | Assert function should print message to stderr at host | Must have |
+| 3 | Stop under debugger | When debugger is attached, break at assertion point | Highly desired |
+| 4 | Reliability | Assert failure should be reported regardless of kernel deadlock | Highly desired |
+
+Implementations without enough capabilities to implement fourth requirement are
+allowed to realize the fallback approach described below, which does not
+guarantee assertion failure delivery to host, but is still useful in many
+practical cases.
+
+
+## Terms
+
+ - Device-side Runtime - runtime library supplied by the Native Device Compiler
+   and running on the device.
+ - Native Device Compiler - compiler which generates device-native binary image
+   based on input SPIR-V image.
+ - Low-level Runtime - the backend/runtime behind DPCPP Runtime attached via the
+   Plugin Interface.
+
+
+## How it works?
+
+`assert(expr)` macro ends up in call to `__devicelib_assert_fail`. This function
+is part of [Device library extension](extensions/C-CXX-StandardLibrary/DeviceLibExtensions.rst#cl_intel_devicelib_cassert).
+
+The format of the assert message is unspecified, but it will always include the
+text of the failing expression, the values of the standard macros `__FILE__` and
+`__LINE__`, and the value of the standard variable `__func__`. If the failing
+assert comes from an `nd_range` `parallel_for` it will also include the global
+ID and the local ID of the failing work item.
```

Comment:
If you have an assertion that occurs in a device function that is called from multiple different kernels, some which are parallel_for and some which are not, how can you know when to put in global and local ID in string?

Reply:
In fallback implementation, the way assert message gets reported to host remains unchanged. For the time being, local/global IDs are put to string with help of SPIR-V built-ins.
For safe implementation, its responsibility of device-side runtime to choose if local/global IDs are needed and to get proper values.

Reply:
How can the device-side runtime know/choose if local/global IDs are needed?

Reply:
Sorry, my bad.
> In fallback implementation, the way assert message gets reported to host remains unchanged. For the time being, local/global IDs are put to string with help of SPIR-V built-ins.
For safe implementation, it's responsibility of device-side runtime to choose if local/global IDs are needed and to get proper values.

The last sentence is wrong here as global and local IDs are fetched by wrapper in `libsycl-crt` regardless of safe/fallback implementation use.
One can see the wrapper's implementation here: https://github.com/intel/llvm/blob/sycl/libdevice/crt_wrapper.cpp#L43